src/thor/costmatrix.cc

```diff
@@ -260,559 +238,528 @@ void CostMatrix::SourceToTarget(Api& request,
+  const auto& hlimits = costing_->GetHierarchyLimits();
+  ignore_hierarchy_limits_ =
+      std::all_of(hlimits.begin() + 1, hlimits.begin() + TileHierarchy::levels().size(),
+                  [](const HierarchyLimits& limits) {
+                    return limits.max_up_transitions == kUnlimitedTransitions;
+                  });
+
+  // Add initial sources status
+  for (const auto exp_dir : {MATRIX_FORW, MATRIX_REV}) {
+    const auto count = locs_count_[exp_dir];
+    locs_status_[exp_dir].reserve(count);
+    hierarchy_limits_[exp_dir].resize(count);
+    adjacency_[exp_dir].resize(count);
+    edgestatus_[exp_dir].resize(count);
+    edgelabel_[exp_dir].resize(count);
+    for (uint32_t i = 0; i < count; i++) {
+      // Allocate the adjacency list and hierarchy limits for this source.
+      // Use the cost threshold to size the adjacency list.
+      edgelabel_[exp_dir][i].reserve(max_reserved_labels_count_);
+      adjacency_[exp_dir][i].reuse(0, current_cost_threshold_, costing_->UnitSize(),
+                                   &edgelabel_[exp_dir][i]);
+      locs_status_[exp_dir].emplace_back(kMaxThreshold);
+      hierarchy_limits_[exp_dir][i] = hlimits;
+    }
   }
 
   // Initialize best connection
   GraphId empty;
   Cost trivial_cost(0.0f, 0.0f);
   Cost max_cost(kMaxCost, kMaxCost);
-  best_connection_.reserve(source_count_ * target_count_);
-  for (uint32_t i = 0; i < source_count_; i++) {
-    for (uint32_t j = 0; j < target_count_; j++) {
+  best_connection_.reserve(locs_count_[MATRIX_FORW] * locs_count_[MATRIX_REV]);
+  for (uint32_t i = 0; i < locs_count_[MATRIX_FORW]; i++) {
+    for (uint32_t j = 0; j < locs_count_[MATRIX_REV]; j++) {
       if (equals(source_locations.Get(i).ll(), target_locations.Get(j).ll())) {
         best_connection_.emplace_back(empty, empty, trivial_cost, 0.0f);
         best_connection_.back().found = true;
       } else {
         best_connection_.emplace_back(empty, empty, max_cost, static_cast<uint32_t>(kMaxCost));
-        source_status_[i].remaining_locations.insert(j);
-        target_status_[j].remaining_locations.insert(i);
+        locs_status_[MATRIX_FORW][i].unfound_connections.insert(j);
+        locs_status_[MATRIX_REV][j].unfound_connections.insert(i);
       }
     }
   }
 
   // Set the remaining number of sources and targets
-  remaining_sources_ = 0;
-  for (const auto& s : source_status_) {
-    if (!s.remaining_locations.empty()) {
-      remaining_sources_++;
+  locs_remaining_[MATRIX_FORW] = 0;
+  for (const auto& s : locs_status_[MATRIX_FORW]) {
+    if (!s.unfound_connections.empty()) {
+      locs_remaining_[MATRIX_FORW]++;
     }
   }
-  remaining_targets_ = 0;
-  for (const auto& t : target_status_) {
-    if (!t.remaining_locations.empty()) {
-      remaining_targets_++;
+  locs_remaining_[MATRIX_REV] = 0;
+  for (const auto& t : locs_status_[MATRIX_REV]) {
+    if (!t.unfound_connections.empty()) {
+      locs_remaining_[MATRIX_REV]++;
```

Comment:
up to here it's mostly variable renaming, hardly any logic change

Reply:
haha ah ok :pray: